### PR TITLE
add SyncQueue and AsyncQueue Protocols

### PR DESCRIPTION
**Pull request by rbuffat**
## What do these changes do?

Add protocols to avoid the use of private classes _SyncQueueProxy and _AsyncQueueProxy when adding type hints. 

## Are there changes in behavior for the user?

I assume that not. But it would be good if somebody more experienced could check this.

## Related issue number

See https://github.com/aio-libs/janus/issues/372

This PR is not intended to be directly merged, but to open a discussion about this approach. 

Using this approach, the following seems to work for adding type hints:

```
import asyncio
import janus


def threaded(sync_q: janus.SyncQueue[int]):
    for i in range(100):
        sync_q.put(i)
    sync_q.join()


async def async_coro(async_q: janus.AsyncQueue[int]):
    for i in range(100):
        val = await async_q.get()
        assert val == i
        async_q.task_done()


async def main():
    queue: janus.Queue[int] = janus.Queue()
    loop = asyncio.get_running_loop()
    fut = loop.run_in_executor(None, threaded, queue.sync_q)
    await async_coro(queue.async_q)
    await fut
    queue.close()
    await queue.wait_closed()


asyncio.run(main())

```

Full disclaimer: I never used Protocols before. I skimmed https://www.python.org/dev/peps/pep-0544 and watched https://www.youtube.com/watch?v=kDDCKwP7QgQ 
Also, I only recently started using typing with Python. 

While this approach seems to work, it looks rather verbose.

### CODE REVIEW

**Comment by rbuffat** [comment]
> The Protocol base class is provided in the typing_extensions package for Python 2.7 and 3.4-3.7. Starting with Python 3.8, Protocol is included in the typing module.


https://mypy.readthedocs.io/en/stable/protocols.html

File: janus/__init__.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 from heapq import heappop, heappush
 from queue import Empty as SyncQueueEmpty
 from queue import Full as SyncQueueFull
-from typing import Any, Callable, Deque, Generic, List, Optional, Set, TypeVar
+from typing import Any, Callable, Deque, Generic, List, Optional, Protocol, Set, TypeVar
```
Comment (asvetlov) [inline janus/__init__.py:10]: Please import `Protocol` from `typing_extensions` module; add `typing-extensions>=3.7.4.3` [here](https://github.com/aio-libs/janus/blob/master/setup.cfg#L50)

**Review by asvetlov — APPROVED** [review]
The proposed solution is a little verbose but technically should work perfectly fine.

Please apply a note about `typing_extensions` before the PR merging.

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks!

**Comment by asvetlov** [comment]
0.7.0 released

**Pull request merged**